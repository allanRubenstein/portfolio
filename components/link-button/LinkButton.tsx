import styled, { css } from 'styled-components';
import { LinkButtonProps } from './types';
import Link from 'next/link';

const LinkButton: React.FunctionComponent<LinkButtonProps> = (
  props: LinkButtonProps,
): JSX.Element => {
  // if href is passed, use Link, else button
  if (props.href) {
    return (
      // render a link with an href
      <Link href={props.href} passHref>
        <LinkWrapper>{props.children}</LinkWrapper>
      </Link>
    );
  } else {
    return (
      // render a button
      <ButtonWrapper onClick={props.onClick}>{props.children}</ButtonWrapper>
    );
  }
};
export default LinkButton;

// shared css for button and link
const LinkAndButtonCss = css`
  /* Base styles */
  color: red;
`;

const LinkWrapper = styled.a`
  ${LinkAndButtonCss}/* link specific styles */
`;
const ButtonWrapper = styled.button`
  ${LinkAndButtonCss}/* button specific styles */
`;
