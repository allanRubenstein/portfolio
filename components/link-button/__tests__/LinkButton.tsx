import React from 'react';
import LinkButton from '../LinkButton';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

test('works', () => {
  // it works :)
  const { container } = render(<LinkButton>test</LinkButton>);

  expect(container).toMatchInlineSnapshot(`
    .c0 {
      color: red;
    }

    <div>
      <button
        class="c0"
      >
        test
      </button>
    </div>
  `);
});

test('a11y', async () => {
  const { container } = render(<LinkButton>test</LinkButton>);
  const results = await axe(container);
  expect(results).toHaveNoViolations();
});
