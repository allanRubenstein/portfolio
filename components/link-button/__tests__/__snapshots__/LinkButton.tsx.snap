// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button snapshot 1`] = `
.c1 {
  line-height: 1.2;
  font-size: 2rem;
  text-transform: uppercase;
  font-family: 'Roboto Condensed',sans-serif;
  font-weight: 700;
  color: var(--white);
}

.c0 {
  position: relative;
  border: none;
  background-color: var(--red);
  box-shadow: 0 4px 0 0 var(--red-dark);
  display: block;
  padding: 1rem 2rem 0.8rem;
  -webkit-transition: background-color var(--transition-short);
  transition: background-color var(--transition-short);
}

.c0:hover {
  background-color: var(--red-light);
  cursor: pointer;
}

.c0:active {
  box-shadow: 0 2px 0 0 var(--red-dark);
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

@media (min-width:600px) {

}

<div>
  <button
    class="c0"
    type="submit"
  >
    <span
      class="c1"
    >
      test
    </span>
  </button>
</div>
`;

exports[`button snapshot 2`] = `
<div>
  <button
    class="LinkButton__ButtonWrapper-sc-4kilkf-1 jmGtyE"
    type="button"
  >
    <span
      class="Title-xde5hf-0 inVoLY"
    >
      test
    </span>
  </button>
</div>
`;

exports[`button snapshot 3`] = `
<div>
  <button
    class="LinkButton__ButtonWrapper-sc-4kilkf-1 faRjjF"
    type="button"
  >
    .c1 {
  position: absolute;
  top: 0px;
  bottom: 0px;
  right: 0px;
  background: var(--red-dark);
  -webkit-transform-origin: 100% 0%;
  -ms-transform-origin: 100% 0%;
  transform-origin: 100% 0%;
  width: .3rem;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,.3rem);
  -ms-transform: translate(.3rem,.3rem);
  transform: translate(.3rem,.3rem);
}

.c2:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleX(0);
  -ms-transform: translate(0rem,0rem) scaleX(0);
  transform: translate(0rem,0rem) scaleX(0);
}

.LinkButton__LinkWrapper-sc-4kilkf-0:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleX(0);
  -ms-transform: translate(0rem,0rem) scaleX(0);
  transform: translate(0rem,0rem) scaleX(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  bottom: 0px;
  left: 0px;
  width: 0px;
  height: 0px;
  border-left: .3rem solid transparent;
  border-right: .3rem solid transparent;
  border-top: .3rem solid var(--red-dark);
  -webkit-transform-origin: center bottom;
  -ms-transform-origin: center bottom;
  transform-origin: center bottom;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(0rem,.3rem) scale(1);
  -ms-transform: translate(0rem,.3rem) scale(1);
  transform: translate(0rem,.3rem) scale(1);
}

.c2:active .c0 {
  -webkit-transform: translate(-.3rem,0rem) scale(0);
  -ms-transform: translate(-.3rem,0rem) scale(0);
  transform: translate(-.3rem,0rem) scale(0);
}

.LinkButton__LinkWrapper-sc-4kilkf-0:active .c0 {
  -webkit-transform: translate(-.3rem,0rem) scale(0);
  -ms-transform: translate(-.3rem,0rem) scale(0);
  transform: translate(-.3rem,0rem) scale(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  bottom: 0px;
  left: 0px;
  right: 0px;
  background: var(--red-dark);
  -webkit-transform-origin: 0px 100%;
  -ms-transform-origin: 0px 100%;
  transform-origin: 0px 100%;
  height: .3rem;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,.3rem) scaleY(1);
  -ms-transform: translate(.3rem,.3rem) scaleY(1);
  transform: translate(.3rem,.3rem) scaleY(1);
}

.c2:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleY(0);
  -ms-transform: translate(0rem,0rem) scaleY(0);
  transform: translate(0rem,0rem) scaleY(0);
}

.LinkButton__LinkWrapper-sc-4kilkf-0:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleY(0);
  -ms-transform: translate(0rem,0rem) scaleY(0);
  transform: translate(0rem,0rem) scaleY(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  top: 0px;
  right: 0px;
  width: 0px;
  height: 0px;
  border-left: 0rem solid transparent;
  border-right: .3rem solid transparent;
  border-bottom: .3rem solid var(--red-dark);
  -webkit-transform-origin: right bottom;
  -ms-transform-origin: right bottom;
  transform-origin: right bottom;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,0rem) scale(1);
  -ms-transform: translate(.3rem,0rem) scale(1);
  transform: translate(.3rem,0rem) scale(1);
}

.c2:active .c0 {
  -webkit-transform: translate(0rem,-.3rem) scale(0);
  -ms-transform: translate(0rem,-.3rem) scale(0);
  transform: translate(0rem,-.3rem) scale(0);
}

.LinkButton__LinkWrapper-sc-4kilkf-0:active .c0 {
  -webkit-transform: translateY(-.3rem) scale(0);
  -ms-transform: translateY(-.3rem) scale(0);
  transform: translateY(-.3rem) scale(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    <span
      class="Title-xde5hf-0 inVoLY"
    >
      test
    </span>
  </button>
</div>
`;

exports[`link snapshot 1`] = `
.c1 {
  line-height: 1.2;
  font-size: 2rem;
  text-transform: uppercase;
  font-family: 'Roboto Condensed',sans-serif;
  font-weight: 700;
  color: var(--white);
}

.c0 {
  position: relative;
  border: none;
  background-color: var(--red);
  box-shadow: 0 4px 0 0 var(--red-dark);
  display: block;
  padding: 1rem 2rem 0.8rem;
  -webkit-transition: background-color var(--transition-short);
  transition: background-color var(--transition-short);
}

.c0:hover {
  background-color: var(--red-light);
  cursor: pointer;
}

.c0:active {
  box-shadow: 0 2px 0 0 var(--red-dark);
  -webkit-transform: translateY(2px);
  -ms-transform: translateY(2px);
  transform: translateY(2px);
}

@media (min-width:600px) {

}

<div>
  <a
    class="c0"
    href="/#"
  >
    <span
      class="c1"
    >
      test
    </span>
  </a>
</div>
`;

exports[`link snapshot 2`] = `
<div>
  <a
    class="LinkButton__LinkWrapper-sc-4kilkf-0 hJmzbW"
    href="/#test2"
  >
    .c1 {
  position: absolute;
  top: 0px;
  bottom: 0px;
  right: 0px;
  background: var(--red-dark);
  -webkit-transform-origin: 100% 0%;
  -ms-transform-origin: 100% 0%;
  transform-origin: 100% 0%;
  width: .3rem;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,.3rem);
  -ms-transform: translate(.3rem,.3rem);
  transform: translate(.3rem,.3rem);
}

.LinkButton__ButtonWrapper-sc-4kilkf-1:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleX(0);
  -ms-transform: translate(0rem,0rem) scaleX(0);
  transform: translate(0rem,0rem) scaleX(0);
}

.c2:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleX(0);
  -ms-transform: translate(0rem,0rem) scaleX(0);
  transform: translate(0rem,0rem) scaleX(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  bottom: 0px;
  left: 0px;
  width: 0px;
  height: 0px;
  border-left: .3rem solid transparent;
  border-right: .3rem solid transparent;
  border-top: .3rem solid var(--red-dark);
  -webkit-transform-origin: center bottom;
  -ms-transform-origin: center bottom;
  transform-origin: center bottom;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(0rem,.3rem) scale(1);
  -ms-transform: translate(0rem,.3rem) scale(1);
  transform: translate(0rem,.3rem) scale(1);
}

.LinkButton__ButtonWrapper-sc-4kilkf-1:active .c0 {
  -webkit-transform: translate(-.3rem,0rem) scale(0);
  -ms-transform: translate(-.3rem,0rem) scale(0);
  transform: translate(-.3rem,0rem) scale(0);
}

.c2:active .c0 {
  -webkit-transform: translate(-.3rem,0rem) scale(0);
  -ms-transform: translate(-.3rem,0rem) scale(0);
  transform: translate(-.3rem,0rem) scale(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  bottom: 0px;
  left: 0px;
  right: 0px;
  background: var(--red-dark);
  -webkit-transform-origin: 0px 100%;
  -ms-transform-origin: 0px 100%;
  transform-origin: 0px 100%;
  height: .3rem;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,.3rem) scaleY(1);
  -ms-transform: translate(.3rem,.3rem) scaleY(1);
  transform: translate(.3rem,.3rem) scaleY(1);
}

.LinkButton__ButtonWrapper-sc-4kilkf-1:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleY(0);
  -ms-transform: translate(0rem,0rem) scaleY(0);
  transform: translate(0rem,0rem) scaleY(0);
}

.c2:active .c0 {
  -webkit-transform: translate(0rem,0rem) scaleY(0);
  -ms-transform: translate(0rem,0rem) scaleY(0);
  transform: translate(0rem,0rem) scaleY(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    .c1 {
  position: absolute;
  top: 0px;
  right: 0px;
  width: 0px;
  height: 0px;
  border-left: 0rem solid transparent;
  border-right: .3rem solid transparent;
  border-bottom: .3rem solid var(--red-dark);
  -webkit-transform-origin: right bottom;
  -ms-transform-origin: right bottom;
  transform-origin: right bottom;
  -webkit-transition: var(--transition-short);
  transition: var(--transition-short);
  -webkit-transform: translate(.3rem,0rem) scale(1);
  -ms-transform: translate(.3rem,0rem) scale(1);
  transform: translate(.3rem,0rem) scale(1);
}

.LinkButton__ButtonWrapper-sc-4kilkf-1:active .c0 {
  -webkit-transform: translate(0rem,-.3rem) scale(0);
  -ms-transform: translate(0rem,-.3rem) scale(0);
  transform: translate(0rem,-.3rem) scale(0);
}

.c2:active .c0 {
  -webkit-transform: translateY(-.3rem) scale(0);
  -ms-transform: translateY(-.3rem) scale(0);
  transform: translateY(-.3rem) scale(0);
}

@media (min-width:600px) {

}

<span
      class="c0 c1"
    />
    <span
      class="Title-xde5hf-0 inVoLY"
    >
      test 2
    </span>
  </a>
</div>
`;
